### PR TITLE
PR: IntuneAppProtectionPolicyiOS - rework 

### LINKED ISSUE #2453: IntuneAppProtectionPolicyAndroid/iOS - Errors when pushing test policies
#### Details of the scenario you tried and the problem that is occurring
Performing testing of Microsoft365DSC, able to export components referenced in subject, compile and create .mof file without much issue. However, when pushing the configuration most of the push seems to run fine until the end which generates several errors. The end result is neither policy is being configured. Given my relatively new experience level, may be simple issue but wanted to post for support.

#### Verbose logs showing the problem
image
VERBOSE: [...-6143KL3]:                            [[IntuneAppProtectionPolicyiOS]da4bdc60-9c8a-4e86-8a35-fd48eaa9fb6d] POST https://graph.microsoft.com/beta/deviceAppManagement/managedAppPolicies
HTTP/1.1 400 Bad Request
Transfer-Encoding: chunked
Vary: Accept-Encoding
Strict-Transport-Security: max-age=31536000
request-id: d6af9341-8c33-4c8e-b37b-0bc45766feee
client-request-id: d6af9341-8c33-4c8e-b37b-0bc45766feee
x-ms-ags-diagnostic: {"ServerInfo":{"DataCenter":"South Central US","Slice":"E","Ring":"5","ScaleUnit":"000","RoleInstance":"SN1PEPF000051D0"}}
Date: Thu, 27 Oct 2022 02:15:38 GMT
Content-Encoding: gzip
Content-Type: application/json

{"error":{"code":"BadRequest","message":"Unable to read JSON request payload. Please ensure Content-Type header is set and payload is of valid JSON format.","innerError":{"date":"2022-10-27T02:15:38","request-id":"d6af9341-8c33-4c8e-b37b-0bc45766feee","clien
t-request-id":"d6af9341-8c33-4c8e-b37b-0bc45766feee"}}}

VERBOSE: [...-6143KL3]:                            [[IntuneAppProtectionPolicyiOS]da4bdc60-9c8a-4e86-8a35-fd48eaa9fb6d] Found Policy with Id {}
Cannot bind argument to parameter 'PolicyId' because it is an empty string.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Set-M365DSCIntuneAppProtectionPolicyiOSAssignment
    + PSComputerName        : localhost
 
VERBOSE: [...-6143KL3]: LCM:  [ End    Set      ]  [[IntuneAppProtectionPolicyiOS]da4bdc60-9c8a-4e86-8a35-fd48eaa9fb6d]  in 15.6690 seconds.
The PowerShell DSC resource '[IntuneAppProtectionPolicyiOS]da4bdc60-9c8a-4e86-8a35-fd48eaa9fb6d' with SourceInfo 'C:\...\M365DSC\Test2ConfigFolder\Compiled_Pushed\M365TenantConfig.ps1::74::9::IntuneAppProtectionPolicyiOS' threw one or more non-terminating 
errors while running the Set-TargetResource functionality. These errors are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
 
VERBOSE: [...-6143KL3]: LCM:  [ End    Set      ]
The SendConfigurationApply function did not succeed.
    + CategoryInfo          : NotSpecified: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : MI RESULT 1
    + PSComputerName        : localhost
 
#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
Local Windows 10 laptop via PowerShell

# insert configuration here
Attached .ps1, .psd1 and .mof files for reference
[M365TenantConfig.zip](https://github.com/microsoft/Microsoft365DSC/files/9875446/M365TenantConfig.zip)

#### The operating system the target node is running
OsName               : Microsoft Windows 10 Enterprise
OsOperatingSystemSKU : EnterpriseEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 19041.1.amd64fre.vb_release.191206-1406
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version of the DSC module that was used ('dev' if using current dev branch)
Name                       : Microsoft365DSC
Version                    : 1.22.1019.1
Type                       : Module
Description                : This DSC module is used to configure and monitor Microsoft tenants, including SharePoint Online, Exchange, Teams, etc.
Author                     : Microsoft Corporation
CompanyName                : NikCharleboisPFE DerekSmay
Copyright                  : (c) 2022 Microsoft Corporation. All rights reserved.
PublishedDate              : 10/19/2022 5:56:49 PM
InstalledDate              : 10/26/2022 8:00:32 PM
UpdatedDate                : 
LicenseUri                 : https://github.com/Microsoft/Microsoft365DSC/blob/master/LICENSE
ProjectUri                 : https://github.com/Microsoft/Microsoft365DSC
IconUri                    : https://github.com/microsoft/Microsoft365DSC/blob/Dev/Modules/Microsoft365DSC/Dependencies/Images/Logo.png?raw=true

### PR DESCRIPTION
**Pull request by William-Francillette**
#### This Pull Request (PR) fixes the following issues

Partial fix #2453
AppProtectionPolicyiOS fixed + missing properties added
#2474 may fix the other resource mentioned in #2453

### CODE REVIEW

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof [hunk 1]
```diff
@@ -1,7 +1,8 @@
 [ClassVersion("1.0.0.0"), FriendlyName("IntuneAppProtectionPolicyiOS")]
 class MSFT_IntuneAppProtectionPolicyiOS : OMI_BaseResource
 {
-    [Key, Description("Display name of the iOS App Protection Policy.")] String DisplayName;
+    [Write, Description("Identity of the iOS App Protection Policy.")] String Identity;
+    [Write, Description("Display name of the iOS App Protection Policy.")] String DisplayName;
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof:5]: Why are we making DisplayName Optional? I believe the new- cmdlet requires it. We should try to align where possible and always require at least one Key parameter.

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof [hunk 2]
```diff
@@ -28,15 +29,30 @@ class MSFT_IntuneAppProtectionPolicyiOS : OMI_BaseResource
...
     [Write, Description("Versions less than the specified version will block the managed app from accessing company data.")] String MinimumRequiredAppVersion;
+    [Write, Description("Versions less than the specified version will result in warning message on the managed app from accessing company data.")] String MinimumWarningAppVersion;
     [Write, Description("Versions less than the specified version will block the managed app from accessing company data.")] String MinimumRequiredOSVersion;
-    [Write, Description("Versions less than the specified version will block the managed app from accessing company data.")] String MinimumRequiredSdkVersion;
-    [Write, Description("Versions less than the specified version will result in warning message on the managed app")] String MinimumWarningAppVersion;
     [Write, Description("Versions less than the specified version will result in warning message on the managed app from accessing company data.")] String MinimumWarningOSVersion;
-    [Write, Description("Require app data to be encrypted.")] String AppDataEncryptionType;
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof:35]: These will be breaking changes. We either need to wait for April's release to push them, or continue to support them until then, but mark them as deprecated. Deprecating a param means updating the Get not to return them, and printing a warning from the set to let the user know that if they use it, that they are deprecated and will not be updated.

**Review by NikCharlebois — CHANGES_REQUESTED** [review]

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof (hunk 2 above)
Comment (William-Francillette) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof:35]: @NikCharlebois, I haven't removed the parameter but added the values map - I reordered the parameters list to suit the doc
`[Write, Description("Require app data to be encrypted."), Values{"useDeviceSettings","afterDeviceRestart","whenDeviceLockedExceptOpenFiles","whenDeviceLocked"}, ValueMap{"useDeviceSettings","afterDeviceRestart","whenDeviceLockedExceptOpenFiles","whenDeviceLocked"}] String AppDataEncryptionType;` line 44

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof (hunk 1 above)
Comment (William-Francillette) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.schema.mof:5]: I make them optional if the displayName needs to be renamed for use case such as change of naming convention

**Review by NikCharlebois — APPROVED** [review]

**Pull request merged**